### PR TITLE
Add optional ResetFieldsFilterStrategy interface for storage

### PR DESCRIPTION
**Pull request by jpbetz**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:

This proposes a more general way of of declaring reset fields to support https://github.com/kubernetes/kubernetes/pull/128266.

Picks up https://github.com/kubernetes-sigs/structured-merge-diff/pull/265 and provides a new interface optional interfaces for strategies to implement when field resetting cannot reasonably be expressed with a exclude filter:

```go
type ResetFieldsFilterStrategy interface {
	GetResetFieldsFilter() map[fieldpath.APIVersion]fieldpath.Filter
}
```

Example usage:

```go

var PodResizeStrategy = podResizeStrategy{
  resetFieldsFilter: fieldpath.NewIncludeMatcherFilter(
    fieldpath.MakePrefixMatcherOrDie("spec", "containers", fieldpath.MatchAnyPathElement(), "resources")),
    fieldpath.MakePrefixMatcherOrDie("spec", "initContainers", fieldpath.MatchAnyPathElement(), "resources")),
}

func (p podResizeStrategy) GetResetFieldsFilter() {
  return map[fieldpath.APIVersion]*fieldpath.Set{
    "v1": p.resetFieldsFilter,
  }
}
```

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: vendor/sigs.k8s.io/structured-merge-diff/v4/fieldpath/set.go [hunk 1]
```diff
@@ -503,3 +503,27 @@ func (s *SetNodeMap) Leaves() *SetNodeMap {
...
+	}
+	return result
+}
+
+type excludeFilter struct {
+	exeludeSet *Set
+}
+
+func (t excludeFilter) Filter(set *Set) *Set {
+	return set.RecursiveDifference(t.exeludeSet)
```
Comment (thockin) [inline vendor/sigs.k8s.io/structured-merge-diff/v4/fieldpath/set.go:528]: typo on `exelude`

**Review by thockin — COMMENTED** [review]
This is not awful as an escape hatch, IMO.  What I can't tell is if it is going to be prohibitively expensive?

File: staging/src/k8s.io/apiserver/pkg/endpoints/installer.go [hunk 2]
```diff
@@ -685,9 +685,16 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 		reqScope.MetaGroupVersion = *a.group.MetaGroupVersion
 	}
 
-	var resetFields map[fieldpath.APIVersion]*fieldpath.Set
-	if resetFieldsStrategy, isResetFieldsStrategy := storage.(rest.ResetFieldsStrategy); isResetFieldsStrategy {
-		resetFields = resetFieldsStrategy.GetResetFields()
+	var resetFieldsFilter map[fieldpath.APIVersion]fieldpath.Filter
```
[the remark this answers is not in the document]
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:688]: It's worth it here.  The machinery here is super fussy.

Added a detailed comment.

File: vendor/sigs.k8s.io/structured-merge-diff/v4/fieldpath/set.go [hunk 3]
```diff
@@ -503,3 +503,27 @@ func (s *SetNodeMap) Leaves() *SetNodeMap {
 	}
 	return out
 }
+
+type Filter interface {
```
[the remark this answers is not in the document]
Comment (jpbetz) [reply vendor/sigs.k8s.io/structured-merge-diff/v4/fieldpath/set.go:624]: I've fleshed out to godoc for the SMD changes, I won't resync the vendor changes into here now, but you can follow along here: https://github.com/kubernetes-sigs/structured-merge-diff/pull/265

**Comment by jpbetz** [comment]
> This is not awful as an escape hatch, IMO. What I can't tell is if it is going to be prohibitively expensive?

It's not.  We're providing an alternative to this:

https://github.com/kubernetes-sigs/structured-merge-diff/blob/f034ec0f9a297cb014e5b5f7f4b4c9e4545777eb/fieldpath/set.go#L108-L113

But alternative does a nearly identical traversal:

https://github.com/jpbetz/structured-merge-diff/blob/a4f44c957880a69403d25f08491d676c142d15e9/fieldpath/set.go#L224-L246

We're making it possible to request more complicated filters that will require more allocs to execute, but for the expected uses, we'e expecting a very narrow set of field paths to be created, so I'm not particularly worried about it.

**Comment by jpbetz** [comment]
@liggitt @dims 

Need a dep bump approval.

To unblock https://github.com/kubernetes/kubernetes/pull/128266, we needed merged https://github.com/kubernetes-sigs/structured-merge-diff/pull/265 int k/structured-merge-diff.  The code has been merged and this picks it up.

I can get separate approval for the machinery wiring change.

**Comment by dims** [comment]
you will still need @liggitt for API approval. the dep changes look good to me.

/approve
/lgtm

**Comment by dims** [comment]
/assign @saschagrunert 

(for cri-client approval)

**Pull request merged**